### PR TITLE
#142: Recognize non-production docker version numbers

### LINKED ISSUE #142: Recognize non-production docker version numbers
## Situation

Version numbers of non production images don't follow the `<major>.<minor>.<fix>` schema in that they add a suffix like `-alpha`, `-beta`, `-RC1` and so on.

Currently the version number scanner does not accept this kind of versions. It would however be useful, because that allows writing integration tests based on `exasol-testcontainers` at an earlier stage in the development cycle.

## Acceptance Criteria

1. Version scanner accepts version suffixes.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #142.

### CODE REVIEW

File: doc/system_requirements.md [hunk 1]
```diff
@@ -206,16 +206,17 @@ Covers:
 Needs: dsn
 
 #### Shortened Docker Image References
-`req~shortened-docker-image-references~1`
+`req~shortened-docker-image-references~2`
 
+Users can reference an Exasol Docker image image names of the following kind:
```
Comment (tomuben) [inline doc/system_requirements.md:211]: Suggest:
Users can reference an Exasol Docker image names of the following kind:
or
Users can reference an "Exasol Docker image" image names of the following kind:

File: doc/design.md [hunk 2]
```diff
@@ -214,23 +214,24 @@ Needs: impl, itest
...
 
 The `ExasolDockerImageReference` can be constructed with the following forms of shortened image references as parameter:
 
-* `<major>`
-* `<major>.<minor>`
-* `<major>.<minor>.<fix>`
-* `<major>.<minor>.<fix>-d<docker-image-revision>`
+* `<major>[-suffix][-d<docker-image-revision>]`
+* `<major>.<minor>[-suffix][-d<docker-image-revision>]`
+* `<major>.<minor>.<fix>[-suffix][-d<docker-image-revision>]`
```
Comment (tomuben) [inline doc/design.md:223]: What happens if the -suffix starts with -d ?

File: src/main/java/com/exasol/containers/VersionBasedExasolDockerImageReference.java [hunk 3]
```diff
@@ -91,12 +81,29 @@ public boolean hasDockerImageRevision() {
...
     public String toString() {
-        if (!this.hasDockerImageRevision() && (this.major >= 7)) {
-            return EXASOL_DOCKER_IMAGE_ID + ":" + this.major + "." + this.minor + "." + this.fix;
+        if (!this.hasDockerImageRevision() && isExasolSevenOrLater()) {
+            return EXASOL_DOCKER_IMAGE_ID + ":" + this.major + "." + this.minor + "." + this.fix
+                    + (this.suffix == null ? "" : "-" + this.suffix);
+        } else {
+            return EXASOL_DOCKER_IMAGE_ID + ":" + this.major + "." + this.minor + "." + this.fix
+                    + (this.suffix == null ? "" : "-" + this.suffix) + "-d"
+                    + (this.hasDockerImageRevision() ? this.dockerImageRevision : 1);
```
Comment (tomuben) [inline src/main/java/com/exasol/containers/VersionBasedExasolDockerImageReference.java:102]:
```suggestion
       return EXASOL_DOCKER_IMAGE_ID + ":" + this.major + "." + this.minor + "." + this.fix
                    + (this.suffix == null ? "" : "-" + this.suffix) + getDockerImageRevision();
                    ....
       private String getDockerImageRevision() {
           return ((!this.hasDockerImageRevision() && isExasolSevenOrLater()) ? "-d"
                    + (this.hasDockerImageRevision() ? this.dockerImageRevision : 1) : null;
        }
        
```

Also check the if, seems a bit strange that older releases always run into the else branch!

File: doc/design.md (hunk 2 above)
Comment (redcatbear) [reply doc/design.md:223]: Made that clearer in the requirement that the distinguishing factor is that `-d` followed by numbers constitutes the image revision. Also it needs to come last as seen in the spec.

To make sure this works, I added more test cases.

**Review by tomuben — COMMENTED** [review]
Please check my comments.

File: doc/system_requirements.md (hunk 1 above)
Comment (redcatbear) [reply doc/system_requirements.md:211]: Good point. Remove duplication.

File: src/main/java/com/exasol/containers/VersionBasedExasolDockerImageReference.java (hunk 3 above)
Comment (redcatbear) [reply src/main/java/com/exasol/containers/VersionBasedExasolDockerImageReference.java]: Made the code more readable by extracting the construction of the reference parts and moving the version distinction down to the image revision construction, where it now has a localized impact.

**Review by tomuben — APPROVED** [review]

**Pull request merged**